plugins/crd/plugin_impl_crd.go

```diff
@@ -0,0 +1,345 @@
+	"github.com/ligato/cn-infra/health/statuscheck"
+	"github.com/ligato/cn-infra/logging"
+	"github.com/ligato/networkservicemesh/pkg/apis/networkservicemesh.io/v1"
+	client "github.com/ligato/networkservicemesh/pkg/client/clientset/versioned"
+	factory "github.com/ligato/networkservicemesh/pkg/client/informers/externalversions"
+)
+
+// Plugin watches K8s resources and causes all changes to be reflected in the ETCD
+// data store.
+type Plugin struct {
+	Deps
+
+	stopCh chan struct{}
+	wg     sync.WaitGroup
+
+	k8sClientConfig *rest.Config
+	k8sClientset    *kubernetes.Clientset
+	apiclientset    *apiextcs.Clientset
+	crdClient       client.Interface
+
+	StatusMonitor statuscheck.StatusReader
+}
+
+// Deps defines dependencies of netmesh plugin.
+type Deps struct {
+	local.PluginInfraDeps
+	// Kubeconfig with k8s cluster address and access credentials to use.
+	KubeConfig config.PluginConfig
+}
+
+var (
+	// queue is a queue of resources to be processed. It performs exponential
+	// backoff rate limiting, with a minimum retry period of 5 seconds and a
+	// maximum of 1 minute.
+	queueNS  = workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*5, time.Minute))
+	queueNSC = workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*5, time.Minute))
+	queueNSE = workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(time.Second*5, time.Minute))
+
+	// stopCh can be used to stop all the informer, as well as control loops
+	// within the application.
+	stopCh = make(chan struct{})
+
+	// sharedFactory is a shared informer factory that is used a a cache for
+	// items in the API server. It saves each informer listing and watching the
+	// same resources independently of each other, thus providing more up to
+	// date results with less 'effort'
+	sharedFactoryNS  factory.SharedInformerFactory
+	sharedFactoryNSE factory.SharedInformerFactory
+	sharedFactoryNSC factory.SharedInformerFactory
+)
+
+// Init builds K8s client-set based on the supplied kubeconfig and initializes
+// all reflectors.
+func (plugin *Plugin) Init() error {
+	var err error
+	plugin.Log.SetLevel(logging.DebugLevel)
+	plugin.stopCh = make(chan struct{})
+
+	kubeconfig := plugin.KubeConfig.GetConfigName()
+	plugin.Log.WithField("kubeconfig", kubeconfig).Info("Loading kubernetes client config")
+	plugin.k8sClientConfig, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to build kubernetes client config: %s", err)
+	}
+
+	plugin.k8sClientset, err = kubernetes.NewForConfig(plugin.k8sClientConfig)
+	if err != nil {
+		return fmt.Errorf("failed to build kubernetes client: %s", err)
+	}
+
+	return nil
+}
+
+// Create the CRD resource, ignore error if it already exists
+func createCRD(plugin *Plugin, FullCRDName, CRDGroup, CRDVersion, CRDPlural, CRDName string) error {
+	crd := &apiextv1beta1.CustomResourceDefinition{
```

Comment:
Verify what is actually returned ("crd" in this case), we may want to utilize this.